db/memtable.cc

```diff
@@ -416,21 +416,32 @@ InternalIterator* MemTable::NewIterator(const ReadOptions& read_options,
   return new (mem) MemTableIterator(*this, read_options, arena);
 }
 
-FragmentedRangeTombstoneIterator* MemTable::NewRangeTombstoneIterator(
-    const ReadOptions& read_options, SequenceNumber read_seq) {
-  if (read_options.ignore_range_deletions ||
-      is_range_del_table_empty_.load(std::memory_order_relaxed)) {
-    return nullptr;
-  }
+void MemTable::RebuildFragmentedTombstones(const ReadOptions& read_options) {
   auto* unfragmented_iter = new MemTableIterator(
       *this, read_options, nullptr /* arena */, true /* use_range_del_table */);
   if (unfragmented_iter == nullptr) {
-    return nullptr;
+    return;
   }
-  auto fragmented_tombstone_list =
+  fragmented_tombstone_list =
       std::make_shared<FragmentedRangeTombstoneList>(
           std::unique_ptr<InternalIterator>(unfragmented_iter),
           comparator_.comparator);
+}
+
+FragmentedRangeTombstoneIterator* MemTable::NewRangeTombstoneIterator(
+    const ReadOptions& read_options, SequenceNumber read_seq) {
+
+  if (read_options.ignore_range_deletions ||
+      is_range_del_table_empty_.load(std::memory_order_relaxed)) {
+    return nullptr;
+  }
+
+  if (fragmented_tombstone_list == nullptr) {
+    RebuildFragmentedTombstones(read_options);
+    if (fragmented_tombstone_list == nullptr) {
+      return nullptr;
+    }
+  }
```

Comment:
If you take my other comment about invalidating `fragmented_tombstone_list` when adding a range tombstone, you'll need to adjust the above to avoid a race between reconstructing the fragmented tombstones and reading them. Specifically, you'll want to change `RebuildFragmentedTombstones` to return the new fragmented tombstone list and use regardless of whether `fragmented_tombstone_list == nullptr`. The thing to pay attention here are concurrent calls to `NewRangeTombstoneIterator` and concurrent calls to `MemTable::Add` to add a new tombstone. As the code is currently written, I believe concurrent calls to `MemTable::Add` can cause the wrong `fragmented_tombstone_list` to be cached. We could see the following scenario:

1. Thread A calls `MemTable::Add(tombstone1)`
1. Thread A calls `RebuildFragmentedTombstones()`
1. Thread A constructs a fragmented tombstone list, but does not assign it to `fragmented_tombstone_list`
1. Thread B calls `MemTable::Add(tombstone2)`
1. Thread B calls `RebuildFragmentedTombstones()`
1. Thread B constructs a fragmented tombstone list and successfully assigns it to `fragmented_tombstone_list`
1. Thread A completes its assignment to `fragmented_tombstone_list`

At this point, `fragmented_tombstone_list` will be incorrect and not contain the `tombstone2` fragments. Something to figure out is when an assignment to `fragmented_tombstone_list` is safe. Given that the count of range tombstones only increases in a `MemTable`, I think you can use that count as a sequence number associated with the fragmented tombstone list and only perform the assignment when the new count is larger than the previous count. You might need to introduce locking to handle this.

Reply:
Okay done.